modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java

```diff
@@ -5524,58 +5527,82 @@ private void restorePartitionStates(
 
             ExecutorService sysPool = ctx.getSystemExecutorService();
 
-            CountDownLatch completionLatch = new CountDownLatch(forGroups.size());
-
             AtomicReference<SortedSet<T3<Long, Long, GroupPartitionId>>> topPartRef = new AtomicReference<>();
 
+            SortedSet<T3<Long, Long, GroupPartitionId>> topParts =
+                new ConcurrentSkipListSet<>(processedPartitionComparator());
+
             long totalPart = forGroups.stream().mapToLong(grpCtx -> grpCtx.affinity().partitions()).sum();
 
-            for (CacheGroupContext grp : forGroups) {
-                sysPool.execute(() -> {
-                    try {
-                        Map<Integer, Long> processed = grp.offheap().restorePartitionStates(partStates);
+            int poolSize = ctx.config().getSystemThreadPoolSize();
 
-                        totalProcessed.addAndGet(processed.size());
+            List<List<GroupPartitionId>> partIds = new ArrayList<>(poolSize);
```

Comment:
Why splitting into batches is better than just groups?

Reply:
This is implied by the task statement - parallelization based on groups causes ineffective resources utilization when groups strongly differ from each other. Sizes of batches should be equal.